### PR TITLE
Add `indexes_in_progress` field to dataset stats

### PR DESCRIPTION
**Pull request by minhtuev**
## What changes are proposed in this pull request?

Added a field to return the list of dataset indices that are in the progress of being built

## How is this patch tested? If it is not, please explain why.

* Manual testing (✅ )
* Unit test (✅ )

### Is this a user-facing change that should be mentioned in the release notes?

-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
  - Enhanced index information retrieval with the addition of an `include_progress` parameter for more comprehensive reporting.
  - New metrics added to statistics, including `indexes_in_progress` and `index_sizes` for improved dataset index state visibility.

- **Bug Fixes**
  - Improved handling and reporting of index build status, ensuring accurate representation of index states.

- **Tests**
  - Added new tests to validate the behavior of dataset indexing, specifically checking the status of index builds.

### CODE REVIEW

**Comment by minhtuev** [comment]
This field is useful for long-running index creation in large datasets

```
In [1]: import fiftyone as fo
   ...: dataset = fo.load_dataset("oi-v7full-10mm")
   ...: dataset.list_indexes()
Out[1]:
['id',
 'negative_labels.classifications.label',
 'positive_labels.classifications.label']

In [2]: dataset.stats(include_indexes=True)
Out[2]:
{'samples_count': 10000000,
 'samples_bytes': 156539929160,
 'samples_size': '145.8GB',
 'indexes_count': 3,
 'indexes_bytes': 437223424,
 'indexes_size': '417.0MB',
 'index_bytes': {'id': 130613248,
  'negative_labels.classifications.label': 101224448,
  'positive_labels.classifications.label': 205385728},
 'index_sizes': {'id': '124.6MB',
  'negative_labels.classifications.label': '96.5MB',
  'positive_labels.classifications.label': '195.9MB'},
 'indexes_in_progress': [],
 'total_bytes': 156977152584,
 'total_size': '146.2GB'}

In [3]: dataset.drop_index("positive_labels.classifications.label")

In [4]: dataset.stats(include_indexes=True)
Out[4]:
{'samples_count': 10000000,
 'samples_bytes': 156539929160,
 'samples_size': '145.8GB',
 'indexes_count': 2,
 'indexes_bytes': 231837696,
 'indexes_size': '221.1MB',
 'index_bytes': {'id': 130613248,
  'negative_labels.classifications.label': 101224448},
 'index_sizes': {'id': '124.6MB',
  'negative_labels.classifications.label': '96.5MB'},
 'indexes_in_progress': [],
 'total_bytes': 156771766856,
 'total_size': '146.0GB'}

In [5]: dataset.create_index("positive_labels.classifications.label")
<interrupted but index creation in the background>

In [6]: dataset.stats(include_indexes=True)
Out[6]:
{'samples_count': 10000000,
 'samples_bytes': 156539929160,
 'samples_size': '145.8GB',
 'indexes_count': 3,
 'indexes_bytes': 231841792,
 'indexes_size': '221.1MB',
 'index_bytes': {'id': 130613248,
  'negative_labels.classifications.label': 101224448,
  'positive_labels.classifications.label': 4096},
 'index_sizes': {'id': '124.6MB',
  'negative_labels.classifications.label': '96.5MB',
  'positive_labels.classifications.label': '4.0KB'},
 'indexes_in_progress': ['positive_labels.classifications.label'],
 'total_bytes': 156771770952,
 'total_size': '146.0GB'}
```

File: fiftyone/core/collections.py [hunk 1]
```diff
@@ -9051,14 +9051,29 @@ def list_indexes(self):
...
+        self, include_size=False, include_progress=False
+    ):
         """Returns a dictionary of information about the indexes on this
         collection.
 
         See :meth:`pymongo:pymongo.collection.Collection.index_information` for
         details on the structure of this dictionary.
 
         include_size (False): whether to include the size of each index
+        include_progress (False): whether to include the build status of index
```
Comment (minhtuev) [inline fiftyone/core/collections.py:9076]: We can combine both of these flags together so there is a single flag

File: fiftyone/core/collections.py [hunk 2]
```diff
@@ -9069,15 +9084,19 @@ def get_index_information(self, include_size=False):
...
-            )
-            for key, size in cs["indexSizes"].items():
+            sample_cs = self._sample_stats()
+            for key, size in sample_cs["indexSizes"].items():
                 if key in sample_info:
                     sample_info[key]["size"] = size
 
+        if include_progress:
+            sample_cs = sample_cs or self._sample_stats()
+            for key in sample_cs["indexBuilds"]:
```
Comment (minhtuev) [inline fiftyone/core/collections.py:9096]: Currently, we only set `in_progress=True` for the index that is being created and skip it for already built indexes; we can also set this field `in_progress=False` for the built indexes as well.

File: fiftyone/core/collections.py [hunk 3]
```diff
@@ -9051,14 +9051,29 @@ def list_indexes(self):
         """
         return list(self.get_index_information().keys())
 
-    def get_index_information(self, include_size=False):
+    def _sample_stats(self) -> dict:
+        """Returns the collection stats for the sample collection."""
+        conn = foo.get_db_conn()
+        cs = conn.command("collstats", self._dataset._sample_collection_name)
+        return cs
```
Comment (benjaminpkane) [inline fiftyone/core/collections.py:9058]:
```suggestion
        return conn.command("collstats", self._dataset._sample_collection_name)
```

File: fiftyone/core/collections.py [hunk 4]
```diff
@@ -9051,14 +9051,29 @@ def list_indexes(self):
...
+
+    def _frame_stats(self) -> dict:
+        """Returns the collection stats for the frame collection."""
+        conn = foo.get_db_conn()
+        cs = conn.command("collstats", self._dataset._frame_collection_name)
+        return cs
```
Comment (benjaminpkane) [inline fiftyone/core/collections.py:9064]:
```suggestion
        return conn.command("collstats", self._dataset._frame_collection_name)
```

File: fiftyone/core/collections.py (hunk 1 above)
Comment (benjaminpkane) [inline fiftyone/core/collections.py:9076]:
```suggestion
        include_progress (False): whether to include the build status of index
        include_size (False): whether to include the size of each index
```

**Review by benjaminpkane — COMMENTED** [review]
Looks great! I believe documentation outputs may need to be updated. Can sync offline

File: tests/unittests/dataset_tests.py [hunk 5]
```diff
@@ -560,6 +562,34 @@ def test_index_sizes(self):
         for d in info.values():
             self.assertTrue(d.get("size") is not None)
 
+    @drop_datasets
+    def test_index_in_progress(self):
+        gt = fo.Detections(detections=[fo.Detection(label="foo")])
+        sample = fo.Sample(filepath="video.mp4", gt=gt)
+        sample.frames[1] = fo.Frame(gt=gt)
```
Comment (benjaminpkane) [inline tests/unittests/dataset_tests.py:569]: Nice 🚀

File: fiftyone/core/collections.py [hunk 6]
```diff
@@ -9051,13 +9051,26 @@ def list_indexes(self):
         """
         return list(self.get_index_information().keys())
 
-    def get_index_information(self, include_size=False):
+    def _sample_stats(self) -> dict:
+        """Returns the collection stats for the sample collection."""
```
Comment (benjaminpkane) [inline fiftyone/core/collections.py:9055]: Private methods don't "require" docstrings. Could also move to simple utils method

**Review by benjaminpkane — APPROVED** [review]
LGTM

**Review by benjaminpkane — APPROVED** [review]
🥼

**Pull request merged**